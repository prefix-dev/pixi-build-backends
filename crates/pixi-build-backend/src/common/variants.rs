//! Variants trait
//!
use std::collections::BTreeMap;

use miette::IntoDiagnostic;
use rattler_build::{NormalizedKey, recipe::variable::Variable, variant_config::VariantConfig};
use rattler_conda_types::Platform;

use crate::ProjectModel;

/// Return variants for the given project model
pub fn compute_variants<P: ProjectModel>(
    project_model: &P,
    input_variant_configuration: Option<BTreeMap<NormalizedKey, Vec<Variable>>>,
    host_platform: Platform,
) -> miette::Result<Vec<BTreeMap<NormalizedKey, Variable>>> {
    // Create a variant config from the variant configuration in the parameters.
    let variant_config = VariantConfig {
        variants: input_variant_configuration.unwrap_or_default(),
        pin_run_as_build: None,
        zip_keys: None,
    };

    // Determine the variant keys that are used in the recipe.
    let used_variants = project_model.used_variants(Some(host_platform));

    // Determine the combinations of the used variants.
    variant_config
        .combinations(&used_variants, None)
        .into_diagnostic()
}
