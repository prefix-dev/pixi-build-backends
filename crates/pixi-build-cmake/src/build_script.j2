{%- set is_cmd_exe = build_platform == "windows" -%}
{%- macro env(key) -%}
{%- if is_cmd_exe %}{{ "%" ~ key ~ "%" }}{% else %}{{ "$" ~key }}{% endif -%}
{% endmacro -%}

{# - Set up common variables -#}
{%- set build_dir = "build" -%}
{%- set source_dir = env("SRC_DIR") -%}
{%- set library_prefix =  "%LIBRARY_PREFIX%" if build_platform == "windows" else "$PREFIX" -%}

{# Set up default CMake arguments -#}
{%- set cmake_args = [
    env("CMAKE_ARGS"),
    "-GNinja",
    "-S \"" ~ source_dir ~ "\"",
    "-DCMAKE_BUILD_TYPE=Release",
    "-DCMAKE_INSTALL_PREFIX=" ~ library_prefix,
    "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON",
    "-DBUILD_SHARED_LIBS=ON",
]
-%}

{# Add Python executable if available -#}
{%- if has_host_python -%}
{%- set cmake_args = cmake_args + [
    "-DPython_EXECUTABLE=" ~ env("PYTHON")
] -%}
{% endif -%}

{#- Output version information -#}
ninja --version
cmake --version

{# Set up the build directory -#}
{% if is_cmd_exe -%}
if not exist {{ build_dir }} mkdir {{ build_dir }}
{% else -%}
mkdir -p build
{% endif -%}
pushd build

{# Windows -#}
{% if is_cmd_exe -%}
if not exist build.ninja (
    cmake {{ cmake_args | join(" ^\n        ") }}
    @if errorlevel 1 exit 1
)

cmake --build . --target install
@if errorlevel 1 exit 1

{# Non Windows #}
{% else -%}
if [ ! -f "build.ninja" ]; then
    cmake {{ cmake_args | join(" \\\n        ") }}
fi

cmake --build . --target install
{% endif -%}
