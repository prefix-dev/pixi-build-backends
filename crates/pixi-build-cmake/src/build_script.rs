use minijinja::Environment;
use serde::Serialize;

#[derive(Serialize)]
pub struct BuildScriptContext {
    pub build_platform: BuildPlatform,
    pub source_dir: String,
    pub extra_args: Vec<String>,
    /// The package has a host dependency on Python.
    /// This is used to determine if the build script
    /// should include Python-related logic.
    pub has_host_python: bool,
}

#[derive(Copy, Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
#[cfg_attr(test, derive(strum::Display))]
#[cfg_attr(test, strum(serialize_all = "snake_case"))]
pub enum BuildPlatform {
    Windows,
    Unix,
}

impl BuildScriptContext {
    pub fn render(&self) -> Vec<String> {
        let env = Environment::new();
        let template = env
            .template_from_str(include_str!("build_script.j2"))
            .unwrap();
        let rendered = template.render(self).unwrap().to_string();
        rendered.lines().map(|s| s.to_string()).collect()
    }
}

#[cfg(test)]
mod test {
    use rstest::*;

    use super::*;

    #[rstest]
    fn test_build_script(
        #[values(BuildPlatform::Windows, BuildPlatform::Unix)] build_platform: BuildPlatform,
        #[values(true, false)] has_host_python: bool,
        #[values(vec![String::from("test-arg")], vec![])] extra_args: Vec<String>,
    ) {
        let context = BuildScriptContext {
            build_platform,
            source_dir: String::from("my-prefix-dir"),
            extra_args: extra_args.clone(),
            has_host_python,
        };
        let script = context.render();

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!(
            "{}-{}-{}",
            build_platform,
            if has_host_python {
                "python"
            } else {
                "nopython"
            },
            if extra_args.is_empty() {
                "no-extra-args"
            } else {
                "with-extra-args"
            }
        ));
        settings.bind(|| {
            insta::assert_snapshot!(script.join("\n"));
        });
    }
}
