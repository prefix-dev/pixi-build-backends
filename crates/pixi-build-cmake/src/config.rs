use std::path::{Path, PathBuf};

use indexmap::IndexMap;
use pixi_build_backend::generated_recipe::BackendConfig;
use serde::Deserialize;

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct CMakeBackendConfig {
    /// Extra args for CMake invocation
    #[serde(default)]
    pub extra_args: Vec<String>,
    /// Environment Variables
    #[serde(default)]
    pub env: IndexMap<String, String>,
    /// If set, internal state will be logged as files in that directory
    pub debug_dir: Option<PathBuf>,
    /// Extra input globs to include in addition to the default ones
    #[serde(default)]
    pub extra_input_globs: Vec<String>,
}

impl BackendConfig for CMakeBackendConfig {
    fn debug_dir(&self) -> Option<&Path> {
        self.debug_dir.as_deref()
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::CMakeBackendConfig;

    #[test]
    fn test_ensure_deseralize_from_empty() {
        let json_data = json!({});
        serde_json::from_value::<CMakeBackendConfig>(json_data).unwrap();
    }
}
