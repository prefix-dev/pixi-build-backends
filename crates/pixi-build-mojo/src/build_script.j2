{%- set is_cmd_exe = build_platform == "windows" -%}
{%- macro env(key) -%}
{%- if is_cmd_exe %}{{ "%" ~ key ~ "%" }}{% else %}{{ "$" ~key }}{% endif -%}
{% endmacro -%}

{# - Set up common variables -#}
{%- set library_prefix =  "%LIBRARY_PREFIX%" if build_platform == "windows" else "$PREFIX" -%}

mojo --version


{#- Build any binaries -#}
{% if bins %}
	{% for bin in bins %}
		mojo build {{ bin.extra_args | join(" ")  }} {{ bin.path }} -o {{ library_prefix }}/bin/{{ bin.name }}
	{% endfor %}
{% endif %}

{#- Build pkg -#}
{% if pkg %}
	mojo package {{ pkg.extra_args | join(" ") }} {{ pkg.path }} -o {{ library_prefix }}/lib/mojo/{{ pkg.name}}.mojopkg
{% endif %}
