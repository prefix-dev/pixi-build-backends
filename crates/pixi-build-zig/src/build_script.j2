{%- macro env(key) -%}
{%- if is_bash %}{{ "$" ~key }}{% else %}{{ "%" ~ key ~ "%" }}{% endif -%}
{%- endmacro -%}
{%- set install_prefix = env("LIBRARY_PREFIX") if not is_bash else env("PREFIX") -%}

zig build --prefix {{ install_prefix }}{% if extra_args | length > 0 %} {{ extra_args | join(" ") }}{% endif %}
{%- if not is_bash %}

if errorlevel 1 exit 1
{%- endif %}
