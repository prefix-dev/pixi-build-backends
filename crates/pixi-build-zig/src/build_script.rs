use minijinja::Environment;
use serde::Serialize;

#[derive(Serialize)]
pub struct BuildScriptContext {
    /// Any additional args to pass to `zig build`
    pub extra_args: Vec<String>,

    /// The platform that is running the build.
    pub is_bash: bool,
}

impl BuildScriptContext {
    pub fn render(&self) -> String {
        let env = Environment::new();
        let template = env
            .template_from_str(include_str!("build_script.j2"))
            .unwrap();
        template.render(self).unwrap().trim().to_string()
    }
}

#[cfg(test)]
mod test {
    use rstest::*;

    #[rstest]
    fn test_build_script(#[values(true, false)] is_bash: bool) {
        let context = super::BuildScriptContext {
            extra_args: vec![],
            is_bash,
        };
        let script = context.render();

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(if is_bash { "bash" } else { "cmdexe" });
        settings.bind(|| {
            insta::assert_snapshot!(script);
        });
    }

    #[rstest]
    fn test_build_script_with_extra_args(#[values(true, false)] is_bash: bool) {
        let context = super::BuildScriptContext {
            extra_args: vec!["-Doptimize=ReleaseFast".to_string()],
            is_bash,
        };
        let script = context.render();

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(if is_bash { "bash" } else { "cmdexe" });
        settings.bind(|| {
            insta::assert_snapshot!(script);
        });
    }
}
