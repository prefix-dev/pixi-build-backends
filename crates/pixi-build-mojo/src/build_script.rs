use super::config::{MojoBinConfig, MojoPkgConfig};
use minijinja::Environment;
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct BuildScriptContext {
    /// The directory where the source code is located, the manifest root.
    pub source_dir: String,
    /// Any executable artifacts to create.
    pub bins: Option<Vec<MojoBinConfig>>,
    /// Any packages to create.
    pub pkg: Option<MojoPkgConfig>,
}

impl BuildScriptContext {
    pub fn render(&self) -> Vec<String> {
        let env = Environment::new();
        let template = env
            .template_from_str(include_str!("build_script.j2"))
            .unwrap();
        let rendered = template.render(self).unwrap().to_string();
        rendered.lines().map(|s| s.to_string()).collect()
    }
}
