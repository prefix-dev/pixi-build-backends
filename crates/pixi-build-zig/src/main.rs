mod build_script;
mod config;

use build_script::BuildScriptContext;
use config::ZigBackendConfig;
use miette::IntoDiagnostic;
use pixi_build_backend::variants::NormalizedKey;
use pixi_build_backend::{
    generated_recipe::{DefaultMetadataProvider, GenerateRecipe, GeneratedRecipe, PythonParams},
    intermediate_backend::IntermediateBackendInstantiator,
};
use pixi_build_types::ProjectModelV1;
use rattler_conda_types::{MatchSpec, PackageName, Platform};
use recipe_stage0::matchspec::PackageDependency;
use recipe_stage0::recipe::{ConditionalRequirements, Script};
use std::collections::HashSet;
use std::{
    collections::BTreeSet,
    path::{Path, PathBuf},
    sync::Arc,
};

#[derive(Default, Clone)]
pub struct ZigGenerator {}

impl GenerateRecipe for ZigGenerator {
    type Config = ZigBackendConfig;

    fn generate_recipe(
        &self,
        model: &ProjectModelV1,
        config: &Self::Config,
        _manifest_root: PathBuf,
        host_platform: Platform,
        _python_params: Option<PythonParams>,
        _variants: &HashSet<NormalizedKey>,
    ) -> miette::Result<GeneratedRecipe> {
        // Create the recipe using the default metadata provider
        let mut generated_recipe =
            GeneratedRecipe::from_model(model.clone(), &mut DefaultMetadataProvider)
                .into_diagnostic()?;

        let requirements = &mut generated_recipe.recipe.requirements;

        let resolved_requirements = ConditionalRequirements::resolve(
            requirements.build.as_ref(),
            requirements.host.as_ref(),
            requirements.run.as_ref(),
            requirements.run_constraints.as_ref(),
            Some(host_platform),
        );

        // Add zig to build requirements if not already present
        let zig_name = PackageName::new_unchecked("zig");
        if !resolved_requirements.build.contains_key(&zig_name) {
            requirements.build.push(
                PackageDependency::Binary(
                    MatchSpec::from_str("zig", rattler_conda_types::ParseStrictness::Lenient)
                        .into_diagnostic()?,
                )
                .into(),
            );
        }

        let build_script = BuildScriptContext {
            extra_args: config.extra_args.clone(),
            is_bash: !Platform::current().is_windows(),
        }
        .render();

        generated_recipe.recipe.build.script = Script {
            content: build_script,
            env: config.env.clone(),
            ..Default::default()
        };

        Ok(generated_recipe)
    }

    /// Returns the build input globs used by the backend.
    fn extract_input_globs_from_build(
        &self,
        config: &Self::Config,
        _workdir: impl AsRef<Path>,
        _editable: bool,
    ) -> miette::Result<BTreeSet<String>> {
        Ok([
            "**/*.zig",
            // Zig build files
            "build.zig",
            "build.zig.zon",
        ]
        .iter()
        .map(|s| s.to_string())
        .chain(config.extra_input_globs.clone())
        .collect())
    }
}

#[tokio::main]
pub async fn main() {
    if let Err(err) = pixi_build_backend::cli::main(|log| {
        IntermediateBackendInstantiator::<ZigGenerator>::new(log, Arc::default())
    })
    .await
    {
        eprintln!("{err:?}");
        std::process::exit(1);
    }
}

#[cfg(test)]
mod tests {
    use indexmap::IndexMap;

    use super::*;

    #[macro_export]
    macro_rules! project_fixture {
        ($($json:tt)+) => {
            serde_json::from_value::<ProjectModelV1>(
                serde_json::json!($($json)+)
            ).expect("Failed to create TestProjectModel from JSON fixture.")
        };
    }

    #[test]
    fn test_input_globs_includes_extra_globs() {
        let config = ZigBackendConfig {
            extra_input_globs: vec!["custom/*.txt".to_string(), "extra/**/*.zig".to_string()],
            ..Default::default()
        };

        let generator = ZigGenerator::default();

        let result = generator
            .extract_input_globs_from_build(&config, PathBuf::new(), false)
            .unwrap();

        // Verify that all extra globs are included in the result
        for extra_glob in &config.extra_input_globs {
            assert!(
                result.contains(extra_glob),
                "Result should contain extra glob: {extra_glob}"
            );
        }

        // Verify that default globs are still present
        assert!(result.contains("**/*.zig"));
        assert!(result.contains("build.zig"));
        assert!(result.contains("build.zig.zon"));
    }

    #[test]
    fn test_zig_is_in_build_requirements() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let generated_recipe = ZigGenerator::default()
            .generate_recipe(
                &project_model,
                &ZigBackendConfig::default(),
                PathBuf::from("."),
                Platform::Linux64,
                None,
                &HashSet::new(),
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe, {
        ".source[0].path" => "[ ... path ... ]",
        ".build.script" => "[ ... script ... ]",
        }, @r#"
        context: {}
        package:
          name: foobar
          version: 0.1.0
        source: []
        build:
          number: ~
          script: "[ ... script ... ]"
        requirements:
          build:
            - zig
          host: []
          run:
            - boltons
          run_constraints: []
        tests: []
        about:
          homepage: ~
          license: ~
          license_file: ~
          summary: ~
          description: ~
          documentation: ~
          repository: ~
        extra: ~
        "#);
    }

    #[test]
    fn test_zig_is_not_added_if_already_present() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    },
                    "buildDependencies": {
                        "zig": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let generated_recipe = ZigGenerator::default()
            .generate_recipe(
                &project_model,
                &ZigBackendConfig::default(),
                PathBuf::from("."),
                Platform::Linux64,
                None,
                &HashSet::new(),
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe, {
        ".source[0].path" => "[ ... path ... ]",
        ".build.script" => "[ ... script ... ]",
        }, @r#"
        context: {}
        package:
          name: foobar
          version: 0.1.0
        source: []
        build:
          number: ~
          script: "[ ... script ... ]"
        requirements:
          build:
            - zig
          host: []
          run:
            - boltons
          run_constraints: []
        tests: []
        about:
          homepage: ~
          license: ~
          license_file: ~
          summary: ~
          description: ~
          documentation: ~
          repository: ~
        extra: ~
        "#);
    }

    #[test]
    fn test_env_vars_are_set() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let env = IndexMap::from([("foo".to_string(), "bar".to_string())]);

        let generated_recipe = ZigGenerator::default()
            .generate_recipe(
                &project_model,
                &ZigBackendConfig {
                    env: env.clone(),
                    ..Default::default()
                },
                PathBuf::from("."),
                Platform::Linux64,
                None,
                &HashSet::new(),
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe.build.script,
        {
            ".content" => "[ ... script ... ]",
        }, @r#"
        content: "[ ... script ... ]"
        env:
          foo: bar
        secrets: []
        "#);
    }
}
