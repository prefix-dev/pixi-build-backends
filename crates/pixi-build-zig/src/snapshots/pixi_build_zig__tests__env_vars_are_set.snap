---
source: crates/pixi-build-zig/src/main.rs
expression: generated_recipe.recipe.build.script
---
content: "[ ... script ... ]"
env:
  foo: bar
secrets: []
